### PR TITLE
arch/arm: Move ARCHCPUFLAGS to Toolchain.defs

### PR DESCRIPTION
**Pull request by no1wudi**
## Summary
Reopen #5637
## Testing
CI

### CODE REVIEW

File: boards/arm/lpc43xx/bambino-200e/README.txt [hunk 1]
```diff
@@ -111,18 +111,6 @@ ports.
...
      CONFIG_ARMV7M_LAZYFPU=y
 
-CFLAGS
-------
-
-Only the recent toolchains have built-in support for the Cortex-M4 FPU.  You will see
-the following lines in each Make.defs file:
-
-  ifeq ($(CONFIG_ARCH_FPU),y)
-    ARCHCPUFLAGS = -mcpu=cortex-m4 -mthumb -march=armv7e-m -mfpu=fpv4-sp-d16 -mfloat-abi=hard
```
Comment (pkarashchenko) [inline boards/arm/lpc43xx/bambino-200e/README.txt:121]: Why is this removed? I do not see that Cortex-M family is touched

**Comment by no1wudi** [comment]
> Also neon flag is dropped. Is this expected?

It's due to the incomplete SoC feature setting since there is a option `CONFIG_ARM_HAVE_NEON`, I'll submit new commit to this PR to keep its origin flags.

File: boards/arm/imx6/sabre-6quad/scripts/Make.defs [hunk 2]
```diff
@@ -34,7 +34,6 @@ ifneq ($(CONFIG_DEBUG_NOOPT),y)
   ARCHOPTIMIZATION += $(MAXOPTIMIZATION) -fno-strict-aliasing
 endif
 
-ARCHCPUFLAGS = -mcpu=cortex-a9 -mfpu=vfpv4-d16 -mthumb
```
[the remark this answers is not in the document]
Comment (no1wudi) [reply boards/arm/imx6/sabre-6quad/scripts/Make.defs:37]: I's only passed to imx6 by https://github.com/apache/incubator-nuttx/commit/de95a8550f8e468396545eab1be11fd33b12be97, it was controlled by CONFIG_ARM_THUMB now in Toolchain.defs

**Comment by pkarashchenko** [comment]
> It's due to the incomplete SoC feature setting since there is a option `CONFIG_ARM_HAVE_NEON`, I'll submit new commit to this PR to keep its origin flags.

Do we expect to have `CONFIG_ARCH_FP_ABI_SOFT` option as well?

File: arch/arm/src/armv7-a/Toolchain.defs [hunk 3]
```diff
@@ -59,18 +59,27 @@ ifeq ($(CONFIG_ARM_THUMB),y)
 ARCHCPUFLAGS += -mthumb
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-a/Toolchain.defs:59]:
```suggestion
  ARCHCPUFLAGS += -mthumb
```

File: boards/arm/phy62xx/phy6222/scripts/Make.defs [hunk 4]
```diff
@@ -53,9 +53,6 @@ NXFLATLDFLAGS1 = -r -Wl,-d -Wl,-warn-common
 NXFLATLDFLAGS2 = $(NXFLATLDFLAGS1) -T$(TOPDIR)/binfmt/libnxflat/gnu-nxflat-pcrel.ld -Wl,-no-check-sections
 LDNXFLATFLAGS = -e main -s 2048
 
-ifneq ($(CROSSDEV),arm-nuttx-elf-)
```
Comment (pkarashchenko) [inline boards/arm/phy62xx/phy6222/scripts/Make.defs:56]: can we merge this change separately ? The CI is failing now on the mainline

Comment (pkarashchenko) [reply boards/arm/phy62xx/phy6222/scripts/Make.defs:56]: Or this is unrelated to CI failure?

Comment (no1wudi) [reply boards/arm/phy62xx/phy6222/scripts/Make.defs:56]: New gcc can only recognize --nostartfile instead of -nostartfiles bellow (they are already deleted from other socs)

Comment (no1wudi) [reply boards/arm/phy62xx/phy6222/scripts/Make.defs:56]: I'll submit it separately soon.

Comment (pkarashchenko) [reply boards/arm/phy62xx/phy6222/scripts/Make.defs:56]: There is
```
ifneq ($(CROSSDEV),sparc-nuttx-elf-)
  LDFLAGS += -nostartfiles -nodefaultlibs
endif
```
that most probably also should be deleted. But let's fix CI first

File: arch/arm/Kconfig [hunk 5]
```diff
@@ -901,6 +901,13 @@ config ARM_HAVE_NEON
 	---help---
 		Decide whether support NEON instruction
 
+config ARM_FP_ABI_SOFT
```
Comment (xiaoxiang781216) [inline arch/arm/Kconfig:904]: why need ARM_FP_ABI_SOFT? If the hardware has FPU, it's better to enable hard, otherwise it must be softfp.

File: arch/arm/src/armv7-a/Toolchain.defs [hunk 6]
```diff
@@ -69,21 +69,29 @@ else ifeq ($(CONFIG_ARCH_CORTEXA9),y)
...
+    ARCHFPUVERSION = vfpv3
   else
-    ARCHCPUFLAGS += -mfpu=vfpv4-d16
+    ARCHFPUVERSION = vfpv4
   endif
+
+  ARCHCPUFLAGS += -mfpu=$(ARCHFPUVERSION)$(ARCHFPUD32FLAGS)
 endif
 
+
```
Comment (xiaoxiang781216) [inline arch/arm/src/armv7-a/Toolchain.defs:98]: remove the blank line

File: arch/arm/Kconfig [hunk 7]
```diff
@@ -908,6 +908,11 @@ config ARM_FP_ABI_SOFT
 	---help---
 		Pass float value via integer register (-mfloat-abi=softfp) 
 
+config ARM_FP_NEON
```
Comment (xiaoxiang781216) [inline arch/arm/Kconfig:911]: why not directly use ARM_HAVE_NEON in Makefile?

File: arch/arm/Kconfig (hunk 5 above)
Comment (pkarashchenko) [reply arch/arm/Kconfig:904]: This may be relevant if the application is compiled with soft and the kernel may use hard. I mean if there are both app and kernel using soft and then kernel is upgraded to a newer version and enable hard, but there is a need to keep same binary for app. Does this makes sense to you?

File: arch/arm/Kconfig (hunk 7 above)
Comment (pkarashchenko) [reply arch/arm/Kconfig:911]: I think because MCU can have neon, but FP usage can be disabled

File: arch/arm/Kconfig (hunk 5 above)
Comment (pkarashchenko) [reply arch/arm/Kconfig:904]: I still check. If none of syscalls use neither float nor double then we can remove softfp option

Comment (pkarashchenko) [reply arch/arm/Kconfig:904]: Also softfp allows to use hard float with toolchain based on GCC 9 and lower.

File: arch/arm/src/armv7-a/Toolchain.defs [hunk 8]
```diff
@@ -56,23 +56,46 @@ endif
...
+  ifeq ($(CONFIG_ARCH_CORTEXA8),y)
+    ARCHFPUVERSION = vfpv3
+  else ifeq ($(CONFIG_ARCH_CORTEXA9),y)
+    ARCHFPUVERSION = vfpv3
+  else
+    ARCHFPUVERSION = vfpv4
+  endif
+
+  ARCHCPUFLAGS += -mfpu=$(ARCHNEONFLAGS)$(ARCHFPUVERSION)$(ARCHFPUD32FLAGS)
```
[the remark this answers is not in the document]
Comment (no1wudi) [reply arch/arm/src/armv7-a/Toolchain.defs:95]: Thanks, I'll fix it

File: arch/arm/Kconfig (hunk 5 above)
Comment (no1wudi) [reply arch/arm/Kconfig:904]: Emm, but it's a bug of old toolchain, I think in this case we should suggest user to update their toolchain instead of do some magic in correct code to make toolchain happy.

And softfp is a valid option (disabled by default), it can make Toolchain.defs more general (if some one want to use softfp, just enable this option instead of modifiy the board level Make.defs)

File: arch/arm/Kconfig (hunk 7 above)
Comment (xiaoxiang781216) [reply arch/arm/Kconfig:911]: This config can't work, FPU and NEON share the same register set. OS save and restore this register set on when FPU is enabled.

File: arch/arm/Kconfig (hunk 5 above)
Comment (xiaoxiang781216) [reply arch/arm/Kconfig:904]: @no1wudi could you try to build some board with fpu and softfp and disable CONFIG_LIBM? I am curious whether the toolchain provide the built-in library(e.g. libgcc.a, libm.a, libsupc++...) with this combination.

File: arch/arm/Kconfig (hunk 7 above)
Comment (no1wudi) [reply arch/arm/Kconfig:911]: image
@xiaoxiang781216 Should we use these recommanded option while NEON exists?

File: arch/arm/Kconfig (hunk 5 above)
Comment (no1wudi) [reply arch/arm/Kconfig:904]: It works on sabre-6quad and giantboard, and see:
```
➜  nuttx git:(rv) ls /usr/lib/gcc/arm-none-eabi/10.3.1/thumb/
nofp/             v7/               v7-a+fp/          v7e-m/            v7e-m+fp/         v7-m/             v7ve+simd/        v8-a/             v8-m.base/        v8-m.main+dp/                                                     
v6-m/             v7-a/             v7-a+simd/        v7e-m+dp/         v7+fp/            v7-r+fp.sp/       v8.1-m.main+mve/  v8-a+simd/        v8-m.main/        v8-m.main+fp/ 
```

Comment (xiaoxiang781216) [reply arch/arm/Kconfig:904]: Thanks for test.

Comment (xiaoxiang781216) [reply arch/arm/Kconfig:904]: ARM_FP_ABI_SOFT->ARM_FPU_ABI_SOFT?

File: arch/arm/Kconfig (hunk 7 above)
Comment (xiaoxiang781216) [reply arch/arm/Kconfig:911]: Yes, I think. fp16/fp32 can get form CONFIG_ARM_HAVE_FPU_D32

Comment (no1wudi) [reply arch/arm/Kconfig:911]: CONFIG_ARM_HAVE_FPU_D32 meas there are 32 float point registers in FPU instead of 32-bit float point.

Comment (xiaoxiang781216) [reply arch/arm/Kconfig:911]:
```
D32 -> 32 double float register -> 64 single float register
       ~~
F32 -> 16 double float register -> 32 single float register
                                   ~~
```

Comment (no1wudi) [reply arch/arm/Kconfig:911]: Got it

File: arch/arm/Kconfig (hunk 5 above)
Comment (no1wudi) [reply arch/arm/Kconfig]: Done

File: arch/arm/src/armv7-a/Toolchain.defs [hunk 9]
```diff
@@ -56,21 +56,51 @@ endif
...
+    ARCHFPUVERSION = vfpv3
+  endif
+
+  # Cortex-A5  | -mfpu=vfpv3-fp16 | -mfpu=vfpv3-d16-fp16 | -mfpu=neon-fp16
+  # Cortex-A7  | -mfpu=vfpv4      | -mfpu=vfpv4-d16      | -mfpu=neon-vfpv4
+  # Cortex-A8  | -mfpu=vfpv3      |                      | -mfpu=neon
+  # Cortex-A9  | -mfpu=vfpv3-fp16 | -mfpu=vfpv3-d16-fp16 | -mfpu=neon-fp16
+  # Cortex-A15 | -mfpu=vfpv4      |                      | -mfpu=neon-vfpv4
+
+  ARCHCPUFLAGS += -mfpu=$(ARCHNEONFLAGS)$(ARCHFPUVERSION)$(ARCHFPUD32FLAGS)$(ARCHFPUFPFLAGS)
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-a/Toolchain.defs:103]: I do not see where `ARCHFPUFPFLAGS` is set

File: arch/arm/src/armv7-a/Toolchain.defs [hunk 10]
```diff
@@ -56,21 +56,51 @@ endif
...
+    ARCHCPUFLAGS += -mfloat-abi=softfp
+  else
+    ARCHCPUFLAGS += -mfloat-abi=hard
+  endif
+  ifeq ($(CONFIG_ARCH_CORTEXA7),y)
+    ifneq ($(CONFIG_ARM_FP_NEON),y)
+      ARCHFPUVERSION = vfpv4
+    endif
+  else
+    ARCHFPUVERSION = vfpv3
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-a/Toolchain.defs:94]:
```suggestion
    ifeq ($(CONFIG_ARM_FP_NEON),y)
      ARCHFPUVERSION =
      ifeq ($(CONFIG_ARCH_CORTEXA8),y)
        ARCHNEONFLAGS = neon
      endif
    else
      ARCHFPUVERSION = vfpv3
    endif
```

File: arch/arm/Kconfig (hunk 5 above)
Comment (pkarashchenko) [reply arch/arm/Kconfig]: > Emm, but it's a bug of old toolchain, I think in this case we should suggest user to update their toolchain instead of do some magic in correct code to make toolchain happy.
> 
> And softfp is a valid option (disabled by default), it can make Toolchain.defs more general (if some one want to use softfp, just enable this option instead of modifiy the board level Make.defs)

From my practice toolchain update is not always a straight forward process and migration to a new toolchain take a lot of time and efforts. So I'm fine if we leave `softfp` as a fallback option for those who still are not ready to change toolchain withing a single day.

File: arch/arm/src/armv7-a/Toolchain.defs (hunk 10 above)
Comment (pkarashchenko) [reply arch/arm/src/armv7-a/Toolchain.defs:94]: Ok. Hold on. I just found info that `-mfpu=neon (alias for neon-vfpv3)`. I will check if `-mfpu=neon-vfpv3` works and get back here

Comment (pkarashchenko) [reply arch/arm/src/armv7-a/Toolchain.defs:94]: Ok. Really GCC docs states that `Note that ‘neon’ is an alias for ‘neon-vfpv3’ and ‘vfp’ is an alias for ‘vfpv2’.`

File: arch/arm/src/armv7-a/Toolchain.defs [hunk 11]
```diff
@@ -56,21 +56,51 @@ endif
...
+    ifneq ($(CONFIG_ARM_NEON),y)
+      ARCHFPUVERSION = vfpv4
+    endif
+  else
+    ARCHFPUVERSION = vfpv3
+  endif
+
+  # Cortex-A5  | -mfpu=vfpv3-fp16 | -mfpu=vfpv3-d16-fp16 | -mfpu=neon-fp16
+  # Cortex-A7  | -mfpu=vfpv4      | -mfpu=vfpv4-d16      | -mfpu=neon-vfpv4
+  # Cortex-A8  | -mfpu=vfpv3      |                      | -mfpu=neon
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-a/Toolchain.defs:99]:
```suggestion
  # Cortex-A8  | -mfpu=vfpv3      |                      | -mfpu=neon (alias for neon-vfpv3)
```

File: arch/arm/src/armv7-m/Toolchain.defs [hunk 12]
```diff
@@ -112,6 +108,16 @@ else # ifeq ($(CONFIG_ARCH_CORTEXM3),y)
...
 endif
 
+ifeq ($(CONFIG_ARCH_FPU),y)
+  ifeq ($(CONFIG_ARM_FPU_ABI_SOFT),y)
+    TOOLCHAIN_MFLOAT += -mfloat-abi=softfp
+  else
+    TOOLCHAIN_MFLOAT += -mfloat-abi=hard
+  endif
+else
+  TOOLCHAIN_MFLOAT = -mfloat-abi=soft
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-m/Toolchain.defs:118]:
```suggestion
  TOOLCHAIN_MFLOAT += -mfloat-abi=soft
```

File: arch/arm/src/armv7-a/Toolchain.defs (hunk 9 above)
Comment (no1wudi) [reply arch/arm/src/armv7-a/Toolchain.defs]: Removed, I's try to support -fp16 flags but find that FP16 extension need more info like ARCH_FPU_FP16

File: arch/arm/src/armv7-a/Toolchain.defs [hunk 13]
```diff
@@ -56,21 +56,61 @@ endif
...
+
+  # Cortex-A5  | -mfpu=vfpv3-fp16 | -mfpu=vfpv3-d16-fp16 | -mfpu=neon-fp16
+  # Cortex-A7  | -mfpu=vfpv4      | -mfpu=vfpv4-d16      | -mfpu=neon-vfpv4
+  # Cortex-A8  | -mfpu=vfpv3      |                      | -mfpu=neon (alias for neon-vfpv3)
+  # Cortex-A9  | -mfpu=vfpv3-fp16 | -mfpu=vfpv3-d16-fp16 | -mfpu=neon-fp16
+  # Cortex-A15 | -mfpu=vfpv4      |                      | -mfpu=neon-vfpv4
+
+  ARCHCPUFLAGS += -mfpu=$(ARCHNEONFLAGS)$(ARCHFPUVERSION)$(ARCHFPUD32FLAGS)
+else
+  ARCHFPUFLAGS += -mfpu=soft
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-a/Toolchain.defs:113]: Screenshot from 2022-03-03 09-18-48

I think it is better to go with `-mfloat-abi=soft` instead

Comment (no1wudi) [reply arch/arm/src/armv7-a/Toolchain.defs:113]: Oh sorry, this is a typo

File: arch/arm/src/armv7-m/Toolchain.defs (hunk 12 above)
Comment (pkarashchenko) [reply arch/arm/src/armv7-m/Toolchain.defs:118]: Or remove from `else # ifeq ($(CONFIG_ARCH_CORTEXM3),y)` case

Comment (pkarashchenko) [reply arch/arm/src/armv7-m/Toolchain.defs:118]: Or set as else for `ifeq ($(CONFIG_ARCH_FPU),y)` in above `ifeq ($(CONFIG_ARCH` selection

File: arch/arm/src/armv7-a/Toolchain.defs (hunk 10 above)
Comment (pkarashchenko) [reply arch/arm/src/armv7-a/Toolchain.defs:95]: @no1wudi we can get back previous logic with less ifs. Sorry, my comment was faster than my searching in GCC docs.

File: arch/arm/src/armv7-a/Toolchain.defs [hunk 14]
```diff
@@ -56,21 +56,61 @@ endif
...
+
+ifeq ($(CONFIG_ARCH_FPU),y)
+  ifeq ($(CONFIG_ARM_FPU_ABI_SOFT),y)
+    ARCHCPUFLAGS += -mfloat-abi=softfp
+  else
+    ARCHCPUFLAGS += -mfloat-abi=hard
+  endif
+
+  ifeq ($(CONFIG_ARCH_CORTEXA7),y)
+    ifneq ($(CONFIG_ARM_NEON),y)
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-a/Toolchain.defs:91]: I think that `vfpv4` should be specified unconditionally. This is according to the table that you provided

File: arch/arm/src/armv7-a/Toolchain.defs (hunk 10 above)
Comment (no1wudi) [reply arch/arm/src/armv7-a/Toolchain.defs:95]: OK

Comment (pkarashchenko) [reply arch/arm/src/armv7-a/Toolchain.defs:95]: Looking into the table that you provided I think the implementation does not match.
We should have something like:
remove `ARCHNEONFLAGS` completely together with
```
ifeq ($(CONFIG_ARM_NEON),y)
  ARCHNEONFLAGS = neon-
  ifneq ($(CONFIG_ARM_HAVE_FPU_D32),y)
    $(error "Number of FPU is alwasys 32 with NEON enabled, please check Kconfig of your SoC.")
  endif
endif
```
change
```
ifeq ($(CONFIG_ARCH_CORTEXA7),y)
  ifeq ($(CONFIG_ARM_NEON),y)
    ARCHFPUVERSION = neon-vfpv4
  else
    ARCHFPUVERSION = vfpv4
  endif
else
  ifeq ($(CONFIG_ARM_NEON),y)
    ARCHFPUVERSION = neon
  else
    ARCHFPUVERSION = vfpv3
  endif
endif

ARCHCPUFLAGS += -mfpu=$(ARCHFPUVERSION)$(ARCHFPUD32FLAGS)
```

Comment (pkarashchenko) [reply arch/arm/src/armv7-a/Toolchain.defs:95]:
```
  ifneq ($(CONFIG_ARM_HAVE_FPU_D32),y)
    $(error "Number of FPU is alwasys 32 with NEON enabled, please check Kconfig of your SoC.")
  endif
```
I think this should be handled by `Kconfig`

Comment (pkarashchenko) [reply arch/arm/src/armv7-a/Toolchain.defs:95]: And possible add `ARCHFPUSUFIX` to be
```
ifeq ($(CONFIG_ARCH_CORTEXA7),y)
  ARCHFPUSUFIX = -fp16
else ifeq ($(CONFIG_ARCH_CORTEXA7),y)
  ARCHFPUSUFIX = -fp16
endif
```
and
```
ARCHCPUFLAGS += -mfpu=$(ARCHFPUVERSION)$(ARCHFPUD32FLAGS)$(ARCHFPUSUFIX)
```

Comment (no1wudi) [reply arch/arm/src/armv7-a/Toolchain.defs:95]: Yes, but not all neon extension support FP16, so we should add a new option ARM_HAVE_NEON_FP16 for it

Comment (no1wudi) [reply arch/arm/src/armv7-a/Toolchain.defs:95]: > ```
>   ifneq ($(CONFIG_ARM_HAVE_FPU_D32),y)
>     $(error "Number of FPU is alwasys 32 with NEON enabled, please check Kconfig of your SoC.")
>   endif
> ```
> 
> I think this should be handled by `Kconfig`

Removed since NEON depend on CONFIG_ARM_HAVE_FPU_D32

**Comment by xiaoxiang781216** [comment]
@no1wudi please squash the change into one.

File: arch/arm/Kconfig [hunk 15]
```diff
@@ -901,6 +901,18 @@ config ARM_HAVE_NEON
...
+	bool "Soft Float ABI"
+	default n
+	depends on ARCH_HAVE_FPU
+	---help---
+		Pass float value via integer register (-mfloat-abi=softfp) 
+
+config ARM_NEON
+	bool "Advanced SIMD (NEON) Extension"
+	default n
+	depends on ARCH_HAVE_FPU && ARM_HAVE_NEON && ARM_HAVE_FPU_D32
```
Comment (xiaoxiang781216) [inline arch/arm/Kconfig:914]: let's ARM_HAVE_FPU_D32 depends on ARCH_HAVE_FPU 
let's ARM_HAVE_NEON depends on ARCH_HAVE_FPU 
let's ARM_NEON depends on ARM_HAVE_NEON

File: arch/arm/src/armv7-a/Toolchain.defs [hunk 16]
```diff
@@ -56,21 +56,52 @@ endif
...
+    ARCHFPUVERSION = neon-vfpv4
+  else
+    ARCHFPUVERSION = vfpv4
+  endif
+else
+  ifeq ($(CONFIG_ARM_NEON),y)
+    ARCHFPUVERSION = neon
+  else
+    ARCHFPUVERSION = vfpv3
+  endif
```
Comment (xiaoxiang781216) [inline arch/arm/src/armv7-a/Toolchain.defs:100]: add two space from line 89 to 100

**Comment by xiaoxiang781216** [comment]
@no1wudi look like that ARM_HAVE_FPU_D32 and ARCH_HAVE_DPFPU is same, should we merge into one?

**Comment by no1wudi** [comment]
@pkarashchenko There is a little change in dependency chain, please review:
* ARM_NEON
  * ARM_HAVE_NEON
    * ARM_HAVE_FPU_D32
      * ARCH_DPFPU
        * ARCH_FPU && ARCH_HAVE_DPFPU

**Comment by no1wudi** [comment]
> I think we should update `config ARCH_CHIP_A1X` and other `ARCH`s that select `select ARCH_CORTEXA` family and add `ARM_HAVE_NEON` and similar where it is relevant

OK

File: arch/arm/Kconfig [hunk 17]
```diff
@@ -889,18 +899,31 @@ config ARM_HAVE_WFE_SEV
 
 config ARM_HAVE_FPU_D32
```
Comment (xiaoxiang781216) [inline arch/arm/Kconfig:900]: change to ARM_HAVE_DPFPU32

File: arch/arm/Kconfig [hunk 18]
```diff
@@ -889,18 +899,31 @@ config ARM_HAVE_WFE_SEV
...
 	bool
-	select ARCH_HAVE_FPU
 	default n
+	depends on ARCH_DPFPU
```
Comment (xiaoxiang781216) [inline arch/arm/Kconfig:903]: select ARCH_HAVE_DPFPU

File: arch/arm/Kconfig [hunk 19]
```diff
@@ -889,18 +899,31 @@ config ARM_HAVE_WFE_SEV
...
 	---help---
-		FPU implemented in the VFPv3-D32 format that supports
+		FPU implemented in the VFPv[3|4]-D32 format that supports
 		32 double-precision floating-point registers.
 
 config ARM_HAVE_NEON
 	bool
 	default n
+	depends on ARM_HAVE_FPU_D32
```
Comment (xiaoxiang781216) [inline arch/arm/Kconfig:911]: select

File: arch/arm/Kconfig [hunk 20]
```diff
@@ -13,7 +13,7 @@ choice
 config ARCH_CHIP_A1X
 	bool "Allwinner A1X"
 	select ARCH_CORTEXA8
-	select ARCH_HAVE_FPU
+	select ARM_HAVE_NEON
```
Comment (xiaoxiang781216) [inline arch/arm/Kconfig:16]: why change

File: arch/arm/Kconfig [hunk 21]
```diff
@@ -82,7 +82,7 @@ config ARCH_CHIP_IMX6
 	select ARCH_CORTEXA9
 	select ARM_THUMB
 	select ARMV7A_HAVE_L2CC_PL310
-	select ARCH_HAVE_FPU
+	select ARM_HAVE_NEON
```
Comment (xiaoxiang781216) [inline arch/arm/Kconfig:85]: why change

**Comment by minabeoki** [comment]
Cortex-A5 has VFPv4
https://developer.arm.com/Processors/Cortex-A5

**Comment by pkarashchenko** [comment]
> Cortex-A5 has VFPv4
> https://developer.arm.com/Processors/Cortex-A5

Yes. That seems to be true. VFPv4 is an optional extension

**Comment by no1wudi** [comment]
Emm , the VFPv3 is from DEN0018A_neon_programmers_guide.pdf  P2-24:
https://developer.arm.com/documentation/den0018/a/

File: arch/arm/Kconfig [hunk 22]
```diff
@@ -887,20 +886,38 @@ config ARM_HAVE_WFE_SEV
...
+
+config ARM_DPFPU32
+	bool "FPU with 32 double-precision register"
+	default y
+	depends on ARCH_DPFPU && ARM_HAVE_DPFPU32
+
+config ARM_NEON
+	bool "Advanced SIMD (NEON) Extension"
+	default y
+	depends on ARM_HAVE_NEON && ARM_DPFPU32
```
[the remark this answers is not in the document]
Comment (no1wudi) [reply arch/arm/Kconfig:919]: Why? NEON is optional while DPFPU32 supported.

File: arch/arm/Kconfig [hunk 23]
```diff
@@ -249,7 +249,6 @@ config ARCH_CHIP_S32K1XX
 config ARCH_CHIP_SAMA5
 	bool "Atmel SAMA5"
 	select ARCH_CORTEXA5
-	select ARCH_HAVE_FPU
```
[the remark this answers is not in the document]
Comment (no1wudi) [reply arch/arm/Kconfig:252]: It's selected in arch/arm/src/sama5/Kconfig by chip model now:
image

File: arch/arm/Kconfig (hunk 20 above)
Comment (no1wudi) [reply arch/arm/Kconfig:16]: A1X support NEON, and ARM_HAVE_NEON will select ALL FPU relative confg.

File: arch/arm/Kconfig [hunk 24]
```diff
@@ -887,20 +886,38 @@ config ARM_HAVE_WFE_SEV
...
+	bool "Soft Float ABI"
+	default n
+	depends on ARCH_HAVE_FPU
+	---help---
+		Pass float value via integer register (-mfloat-abi=softfp) 
+
+config ARM_DPFPU32
+	bool "FPU with 32 double-precision register"
+	default y
+	depends on ARCH_DPFPU && ARM_HAVE_DPFPU32
```
[the remark this answers is not in the document]
Comment (no1wudi) [reply arch/arm/Kconfig:914]: Why ? The relationship with ARM_DPFPU32 and ARCH_DPFPU is like  ARCH_DPFPU and ARCH_FPU， you can enable DPFPU with ARCH_FPU disabled.

File: arch/arm/Kconfig (hunk 21 above)
Comment (no1wudi) [reply arch/arm/Kconfig:85]: See above

File: arch/arm/Kconfig (hunk 22 above)
[the remark this answers is not in the document]
Comment (no1wudi) [reply arch/arm/Kconfig:919]: image

File: arch/arm/Kconfig (hunk 24 above)
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply arch/arm/Kconfig:914]: Ok

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
@minabeoki @no1wudi which spec should we follow? @minabeoki do you have Cortex-A5 device to do some smoke test?

**Comment by no1wudi** [comment]
> @minabeoki @no1wudi which spec should we follow? @minabeoki do you have Cortex-A5 device to do some smoke test?

https://developer.arm.com/documentation/100302/0001/introduction/about-the-cortex-a5-fpu?lang=en

Maybe we should follow this spec since it's for cortex a5 special, I'll update this PR according to it.

**Comment by minabeoki** [comment]
@no1wudi 
Now, I don't have Cortex-A5 device/board.
It seems certain that Cortex-A5 has vfpv4.
https://developer.arm.com/documentation/ddi0450/b/Introduction/About-the-Cortex-A5-NEON-MPE?lang=en

But, the option is vfpv3 in DEN0018A_neon_programmers_guide.pdf.
I don't know why vfpv3 is specified for A5's compile option.

> Maybe we should follow this spec since it's for cortex a5 special, I'll update this PR according to it.

ok, I agree.

**Comment by xiaoxiang781216** [comment]
@minabeoki do you think this patch is ready for merge?

**Review by pkarashchenko — APPROVED** [review]
LGTM!

**Comment by minabeoki** [comment]
@xiaoxiang781216 
I compiled for sabre-6quad(A9) config, and checked the compiled assemble code.
In case of "-mfpu=neon-vfpv3", GCC uese NEON instructions for integer operations.
In case of "-mfpu=vfpv3", GCC does not uses for integer operations.

"-mfpu=neon-vfpv3" option may be dangerous for the kernel context switch mechanism.
What do you think about this ?

For example, procfs_statfs function,
https://github.com/apache/incubator-nuttx/blob/779665c7045b7661ebdf4baf0da8f4a0264795a1/fs/procfs/fs_procfs.c#L997-L1009

"-mfpu=neon-vfpv3" case:
(vmov and vstr are used for 64bit store.)
```
10807d00 <procfs_statfs>:
10807d00:       b510            push    {r4, lr}
10807d02:       460c            mov     r4, r1
10807d04:       2238            movs    r2, #56 ; 0x38
10807d06:       2100            movs    r1, #0
10807d08:       4620            mov     r0, r4
10807d0a:       f7fb f9d8       bl      108030be <memset>
10807d0e:       4b07            ldr     r3, [pc, #28]   ; (10807d2c <procfs_statfs+0x2c>)
10807d10:       2000            movs    r0, #0
10807d12:       efc0 0010       vmov.i32        d16, #0 ; 0x00000000
10807d16:       60a0            str     r0, [r4, #8]
10807d18:       6023            str     r3, [r4, #0]
10807d1a:       2320            movs    r3, #32
10807d1c:       edc4 0b04       vstr    d16, [r4, #16]
10807d20:       edc4 0b06       vstr    d16, [r4, #24]
10807d24:       edc4 0b08       vstr    d16, [r4, #32]
10807d28:       6063            str     r3, [r4, #4]
10807d2a:       bd10            pop     {r4, pc}
10807d2c:       434f5250        .word   0x434f5250
```

"-mfpu=vfpv3" case:
(There are no neon instructions)
```
10807d08 <procfs_statfs>:
10807d08:   b510        push    {r4, lr}
10807d0a:   460c        mov r4, r1
10807d0c:   2238        movs    r2, #56 ; 0x38
10807d0e:   2100        movs    r1, #0
10807d10:   4620        mov r0, r4
10807d12:   f7fb f9d8   bl  108030c6 <memset>
10807d16:   4b07        ldr r3, [pc, #28]   ; (10807d34 <procfs_statfs+0x2c>)
10807d18:   2200        movs    r2, #0
10807d1a:   2000        movs    r0, #0
10807d1c:   60a0        str r0, [r4, #8]
10807d1e:   6023        str r3, [r4, #0]
10807d20:   2300        movs    r3, #0
10807d22:   e9c4 2304   strd    r2, r3, [r4, #16]
10807d26:   e9c4 2306   strd    r2, r3, [r4, #24]
10807d2a:   e9c4 2308   strd    r2, r3, [r4, #32]
10807d2e:   2320        movs    r3, #32
10807d30:   6063        str r3, [r4, #4]
10807d32:   bd10        pop {r4, pc}
10807d34:   434f5250    .word   0x434f5250
```

**Comment by pkarashchenko** [comment]
@minabeoki do you recommend to default NEON to `n`?

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216 I compiled for sabre-6quad(A9) config, and checked the compiled assemble code. In case of "-mfpu=neon-vfpv3", GCC uese NEON instructions for integer operations. In case of "-mfpu=vfpv3", GCC does not uses for integer operations.
> 
> "-mfpu=neon-vfpv3" option may be dangerous for the kernel context switch mechanism. What do you think about this ?
> 

It will break in the current implementation since FPU context doesn't in:
https://github.com/apache/incubator-nuttx/blob/master/arch/arm/src/armv7-a/arm_vectors.S
but delay to:
https://github.com/apache/incubator-nuttx/blob/master/arch/arm/src/armv7-a/arm_savefpu.S
If the compiler generate any code which utilize NEON instruction even without any floating point variables, the interrupted thread context will be destroyed.

@anchao is preparing a patch to optimize FPU save/restore performance which can fix this problem too.

**Comment by xiaoxiang781216** [comment]
> @minabeoki do you recommend to default NEON to `n`?

We need fix this problem instead, the default value just hide/postpone the problem. If we can't fix this problem, NEON option must be removed from Kconfig and Make.defs.

**Comment by pkarashchenko** [comment]
> ill break in the current implementation since FPU context doesn't in

Is disabling NEON for kernel build an option? I mean can we still use NEON for applications?

**Comment by xiaoxiang781216** [comment]
So you want CONFIG_NEON only change the userspace code, this could be worked, but it's better to fix the root cause. Let's wait @anchao 's patch before we merge this one?

**Comment by pkarashchenko** [comment]
> So you want CONFIG_NEON only change the userspace code, this could be worked, but it's better to fix the root cause. Let's wait @anchao 's patch before we merge this one?

Yes. For sure. I agree to wait. I meant that in the worst case we could make a dependency to allow NEON for user space only.

**Review by hartmannathan — COMMENTED** [review]
Does this PR need to be marked with "breaking change" label (requiring manual changes for out-of-tree boards)?

**Comment by xiaoxiang781216** [comment]
> Does this PR need to be marked with "breaking change" label (requiring manual changes for out-of-tree boards)?

Since the same gcc option can be specified multiple times, the out-of-tree boards could keep as before, but it's always better to remove the duplicated option.

**Comment by xiaoxiang781216** [comment]
@no1wudi this patch need rebase to the last mainline.

**Comment by pkarashchenko** [comment]
@no1wudi are all necessary changes to mainline done? Can we move forward with NEON now?

**Comment by xiaoxiang781216** [comment]
> @no1wudi are all necessary changes to mainline done? Can we move forward with NEON now?

We can merge this PR, since this patch just move the compiler options to the common place, but don't change them.

**Comment by xiaoxiang781216** [comment]
@minabeoki NEON issue is fixed here: https://github.com/apache/incubator-nuttx/pull/6072
